mtags/src/main/scala-2/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -68,78 +73,165 @@ trait MatchCaseCompletions { this: MetalsGlobal =>
         }
-      }
 
-      // Step 1: walk through scope members.
-      metalsScopeMembers(pos).iterator
-        .foreach(m => visit(m.sym.dealiased, Identifier(m.sym.name), Nil))
+        // Step 1: walk through scope members.
+        metalsScopeMembers(pos).iterator
+          .foreach { m =>
+            val sym = m.sym.dealiased
+            val fsym = sym.dealiasedSingleType
+            val isValid = !parents.isParent(fsym) &&
+              (fsym.isCase ||
+                fsym.hasModuleFlag ||
+                fsym.isInstanceOf[TypeSymbol]) &&
+              parents.isSubClass(fsym, includeReverse = false)
+            if (isValid) visit(sym, Identifier(m.sym.name), Nil)
+          }
 
-      // Step 2: walk through known direct subclasses of sealed types.
-      val autoImport = autoImportPosition(pos, text)
-      parents.selector.typeSymbol.foreachKnownDirectSubClass { sym =>
-        autoImport match {
-          case Some(value) =>
-            val (shortName, edits) =
-              ShortenedNames.synthesize(sym, pos, context, value)
-            visit(sym, shortName, edits)
-          case scala.None =>
-            visit(sym, sym.fullNameSyntax, Nil)
+        // Step 2: walk through known direct subclasses of sealed types.
+        // In `List(foo).map { cas@@} we want to provide also `case (exhaustive)` completion
+        // which works like exhaustive match, so we need to collect only members from this step
+        val autoImport = autoImportPosition(pos, text)
+        val sealedDescs = mutable.Set.empty[Symbol]
+        selectorSym.foreachKnownDirectSubClass { sym =>
+          sealedDescs += sym.dealiased
+          autoImport match {
+            case Some(value) =>
+              val (shortName, edits) =
+                ShortenedNames.synthesize(sym, pos, context, value)
+              visit(sym, shortName, edits)
+            case None =>
+              visit(sym, sym.fullNameSyntax, Nil)
+          }
         }
+        val members = result.result()
+        val edits = members.map(_._2)
+        // In `List(foo).map { cas@@} we want to provide also `case (exhaustive)` completion
+        // which works like exhaustive match, so we need to collect only members from this step
+        if (includeExhaustive) {
+          def isSealedDesc(sym: Symbol) = {
+            sym.info match {
+              case tr: TypeRef => sealedDescs(tr.underlying.typeSymbol)
+              case _ => sealedDescs(sym)
+            }
+          }
+          val sortedMembers = sortSubclasses(members, selectorSym.tpe, source)
+          val sealedMembers = sortedMembers.collect {
+            case (sym, member) if isSealedDesc(sym) => member
+          }
+          sealedMembers match {
+            case Nil => edits
+            case head :: tail =>
+              val newText = new l.TextEdit(
+                editRange,
+                tail
+                  .map(_.label.getOrElse(""))
```

Comment:
There is an issue when first `case` statement appears not in the same line as selector, e.g.
```scala
List(Option(1)).map{
  case@@
}

//exhaustive case results in
List(Option(1)).map{
  
    case None => 
    case Some(value) =>
    

}
```
But this problem is common for both scala 2 and 3 so maybe we can fix it in a separate pr?

Reply:
Sure, we can do that separately, but is there a test for that?

Reply:
No, but I will work on this issue today and add it

Reply:
ok, thanks!